capsules/src/touch.rs

```diff
@@ -0,0 +1,326 @@
+//! Provides userspace with access to the touch panel.
+//!
+//! Usage
+//! -----
+//!
+//! You need a touch that provides the `hil::touch::Touch` trait.
+//! An optional gesture client and a screen can be connected to it.
+//!
+//! ```rust
+//! let touch =
+//!     components::touch::TouchComponent::new(board_kernel, ts, Some(ts), Some(screen)).finalize(());
+//! ```
+
+// use core::convert::From;
+use core::mem;
+use kernel::debug;
+use kernel::hil;
+use kernel::hil::screen::ScreenRotation;
+use kernel::hil::touch::{GestureEvent, TouchEvent, TouchStatus};
+use kernel::ReturnCode;
+use kernel::{AppId, AppSlice, Callback, Driver, Grant, Shared};
+
+/// Syscall driver number.
+use crate::driver;
+pub const DRIVER_NUM: usize = driver::NUM::Touch as usize;
+
+pub struct App {
+    touch_callback: Option<Callback>,
+    gesture_callback: Option<Callback>,
+    multi_touch_callback: Option<Callback>,
+    events_buffer: Option<AppSlice<Shared, u8>>,
+    ack: bool,
```

Comment:
This avoids a read write race. Apps will get the scheduled callback, but they might be in the middle of it when the touch screen sends a new set of touches. The data would be overwritten by the capsule right when the app reads it. To avoid this, I will add a command so that apps can ack that they have read the data.

Am I maybe wrong about how callbacks work and the ack is not necessary?

Reply:
No you're right, that is something that has to be considered.